ts/Core/Legend/Legend.ts

```diff
@@ -789,15 +789,21 @@ class Legend {
         }
 
         // Always update the text
-        legend.setText(item);
+        legend.setText(item as Legend.Item);
 
         // calculate the positions for the next line
         const bBox = li.getBBox();
 
         item.itemWidth = item.checkboxOffset =
-            options.itemWidth ||
+            (options.itemWidth ||
             item.legendItemWidth ||
-            bBox.width + itemExtraWidth;
+            bBox.width + itemExtraWidth) + (
+                (options.layout === 'vertical') &&
+                (item as ColorAxis).maxLabelLength &&
+                ((item as ColorAxis).maxLabelLength > itemExtraWidth) &&
+                ((item as ColorAxis).maxLabelLength - itemExtraWidth) ||
+                0
```

Comment:
Also to avoid polluting the core with code dealing with ColorAxis, check if this code could be moved to the ColorAxis. The legend is already checking for `item.legendItemWidth`. So it should be possible to make a `legendItemWidth` property on the ColorAxis, and set it there in order for the legend to pick it up.